<h1>Table of Contents<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#Загрузка-и-подготовка-данных" data-toc-modified-id="Загрузка-и-подготовка-данных-1"><span class="toc-item-num">1&nbsp;&nbsp;</span>Загрузка и подготовка данных</a></span></li><li><span><a href="#Разбиение-данных-на-обучающую-и-валидационную-выборки-в-соотношении-75:25" data-toc-modified-id="Разбиение-данных-на-обучающую-и-валидационную-выборки-в-соотношении-75:25-2"><span class="toc-item-num">2&nbsp;&nbsp;</span>Разбиение данных на обучающую и валидационную выборки в соотношении 75:25</a></span></li><li><span><a href="#Обучение-и-проверка-модели" data-toc-modified-id="Обучение-и-проверка-модели-3"><span class="toc-item-num">3&nbsp;&nbsp;</span>Обучение и проверка модели</a></span></li><li><span><a href="#Подготовка-к-расчёту-прибыли" data-toc-modified-id="Подготовка-к-расчёту-прибыли-4"><span class="toc-item-num">4&nbsp;&nbsp;</span>Подготовка к расчёту прибыли</a></span><ul class="toc-item"><li><span><a href="#Рассчет-минимально-достаточного-объём-сырья-для-безубыточной-разработки-новой-скважины." data-toc-modified-id="Рассчет-минимально-достаточного-объём-сырья-для-безубыточной-разработки-новой-скважины.-4.1"><span class="toc-item-num">4.1&nbsp;&nbsp;</span>Рассчет минимально достаточного объём сырья для безубыточной разработки новой скважины.</a></span></li></ul></li><li><span><a href="#Расчёт-прибыли-и-рисков-для-каждого-региона" data-toc-modified-id="Расчёт-прибыли-и-рисков-для-каждого-региона-5"><span class="toc-item-num">5&nbsp;&nbsp;</span>Расчёт прибыли и рисков для каждого региона</a></span><ul class="toc-item"><li><span><a href="#Создадим-функцию-для-расчёта-прибыли-по-выбранным-скважинам-и-предсказаниям-модели" data-toc-modified-id="Создадим-функцию-для-расчёта-прибыли-по-выбранным-скважинам-и-предсказаниям-модели-5.1"><span class="toc-item-num">5.1&nbsp;&nbsp;</span>Создадим функцию для расчёта прибыли по выбранным скважинам и предсказаниям модели</a></span></li><li><span><a href="#Рассчитаем-продуктивность-и-прибыль--по-выбранным-скважинам-и-предсказаниям-модели" data-toc-modified-id="Рассчитаем-продуктивность-и-прибыль--по-выбранным-скважинам-и-предсказаниям-модели-5.2"><span class="toc-item-num">5.2&nbsp;&nbsp;</span>Рассчитаем продуктивность и прибыль  по выбранным скважинам и предсказаниям модели</a></span></li><li><span><a href="#Создадим-функцию-для-расчёта-рисков-и-распределения-прибыли-техникой-Bootstrap" data-toc-modified-id="Создадим-функцию-для-расчёта-рисков-и-распределения-прибыли-техникой-Bootstrap-5.3"><span class="toc-item-num">5.3&nbsp;&nbsp;</span>Создадим функцию для расчёта рисков и распределения прибыли техникой Bootstrap</a></span></li><li><span><a href="#Найдем-среднюю-прибыль,-95%-й-доверительный-интервал-и-риск-убытков-в-трех-регионах" data-toc-modified-id="Найдем-среднюю-прибыль,-95%-й-доверительный-интервал-и-риск-убытков-в-трех-регионах-5.4"><span class="toc-item-num">5.4&nbsp;&nbsp;</span>Найдем среднюю прибыль, 95%-й доверительный интервал и риск убытков в трех регионах</a></span></li></ul></li><li><span><a href="#Общие-выводы" data-toc-modified-id="Общие-выводы-6"><span class="toc-item-num">6&nbsp;&nbsp;</span>Общие выводы</a></span></li></ul></div>

# Выбор локации для скважины

**Цель проекта:**  
Разработать для сотрудников нефтедобывающей компании модель, помогающую принять решение, где бурить новую скважину.

Для проекта есть в наличии пробы нефти в трёх регионах: в каждом 10 000 месторождений, где измерили качество нефти и объём её запасов. 

В рамках проекта будет построена модель машинного обучения с использованием линейной регрессии, которая поможет определить регион, где добыча принесёт наибольшую прибыль. Анализ возможной прибыли и риски будут проанализированы техникой *Bootstrap.*  

Ключевые шаги для выбора локации для бурения новых скважен:
- В избранном регионе ищем месторождения, для каждого определяем значения признаков;
- Строим модель и оцениваем объём запасов;
- Выбираем месторождения с самым высокими оценками значений. 
 - Количество месторождений зависит от бюджета компании и стоимости разработки одной скважины;
- Прибыль равна суммарной прибыли отобранных месторождений.

Дополнительные условия:
- При разведке региона исследуем 500 точек, из которых с помощью машинного обучения выбираем 200 лучших для разработки.
- Бюджет на разработку скважин в регионе — 10 млрд рублей.
- При нынешних ценах один баррель сырья приносит 450 рублей дохода. 
- Доход с каждой единицы продукта составляет 450 тыс. рублей, поскольку объём указан в тысячах баррелей.
- После оценки рисков оставим лишь те регионы, в которых вероятность убытков меньше 2.5%. 
- Среди них выбирем регион с наибольшей средней прибылью.


## Загрузка и подготовка данных

**Импорт библиотек**

In [1]:
import pandas as pd
import numpy as np
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression
from sklearn.metrics import mean_squared_error

**Определение констант для работы в проекте**

In [2]:
# количество точек для разведки в регионе:
EXPLORATION_QUANTITY = 500

# количество выбранных из них лучших для разработки:
BEST_QUANTITY = 200

# бюджет в рублях на разработку скважин в регионе:
DEV_BUDGET = 10e9

# доход в рублях с каждой единицы продукта, которая равна тысяче бареллей:
UNIT_REVENUE = 450_000

# верхний порог вероятности убытков в процентах для выбора региона:
LOSS_PROBABILITY = 2.5

**Определение функции для импорта данных**

In [3]:
def data_import(short_path,long_path):
    
    #импорт данных
    try:
        df = pd.read_csv(short_path)
    except:
        df = pd.read_csv(long_path)
    
    #вывод параметров датасета для проверки    
    print(df.shape)
    print(df.info())
    display(df.head())
    
    #Проверка на явные дубликаты
    print('Количество явных дубликатов', df.duplicated().sum())
    
    return df

**Импорт данных**

In [4]:
# импорт и проверка параметров датасета 0
print('Нефтяное месторождение 0')
df_0 = data_import('geo_data_0.csv',
                   '/datasets/geo_data_0.csv')

# импорт и проверка параметров датасета 1
print('Нефтяное месторождение 1')
df_1 = data_import('geo_data_1.csv',
                   '/datasets/geo_data_1.csv')

# импорт и проверка параметров датасета 2
print('Нефтяное месторождение 2')
df_2 = data_import('geo_data_2.csv',
                   '/datasets/geo_data_2.csv')

Нефтяное месторождение 0
(100000, 5)
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 100000 entries, 0 to 99999
Data columns (total 5 columns):
 #   Column   Non-Null Count   Dtype  
---  ------   --------------   -----  
 0   id       100000 non-null  object 
 1   f0       100000 non-null  float64
 2   f1       100000 non-null  float64
 3   f2       100000 non-null  float64
 4   product  100000 non-null  float64
dtypes: float64(4), object(1)
memory usage: 3.8+ MB
None


id        f0        f1        f2     product
0  txEyH  0.705745 -0.497823  1.221170  105.280062
1  2acmU  1.334711 -0.340164  4.365080   73.037750
2  409Wp  1.022732  0.151990  1.419926   85.265647
3  iJLyR -0.032172  0.139033  2.978566  168.620776
4  Xdl7t  1.988431  0.155413  4.751769  154.036647

Количество явных дубликатов 0
Нефтяное месторождение 1
(100000, 5)
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 100000 entries, 0 to 99999
Data columns (total 5 columns):
 #   Column   Non-Null Count   Dtype  
---  ------   --------------   -----  
 0   id       100000 non-null  object 
 1   f0       100000 non-null  float64
 2   f1       100000 non-null  float64
 3   f2       100000 non-null  float64
 4   product  100000 non-null  float64
dtypes: float64(4), object(1)
memory usage: 3.8+ MB
None


id         f0         f1        f2     product
0  kBEdx -15.001348  -8.276000 -0.005876    3.179103
1  62mP7  14.272088  -3.475083  0.999183   26.953261
2  vyE1P   6.263187  -5.948386  5.001160  134.766305
3  KcrkZ -13.081196 -11.506057  4.999415  137.945408
4  AHL4O  12.702195  -8.147433  5.004363  134.766305

Количество явных дубликатов 0
Нефтяное месторождение 2
(100000, 5)
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 100000 entries, 0 to 99999
Data columns (total 5 columns):
 #   Column   Non-Null Count   Dtype  
---  ------   --------------   -----  
 0   id       100000 non-null  object 
 1   f0       100000 non-null  float64
 2   f1       100000 non-null  float64
 3   f2       100000 non-null  float64
 4   product  100000 non-null  float64
dtypes: float64(4), object(1)
memory usage: 3.8+ MB
None


id        f0        f1        f2     product
0  fwXo0 -1.146987  0.963328 -0.828965   27.758673
1  WJtFt  0.262778  0.269839 -2.530187   56.069697
2  ovLUW  0.194587  0.289035 -5.586433   62.871910
3  q6cA6  2.236060 -0.553760  0.930038  114.572842
4  WPMUX -0.515993  1.716266  5.899011  149.600746

Количество явных дубликатов 0


**Вывод:**  
пропусков и явных дубликатов в данных нет, но при разбиении выборок,  
из обучающих признаков необходимо убрать колонку 'id',  
которая не помогает обучению, но вызывает ошибку в работе модели.

## Разбиение данных на обучающую и валидационную выборки в соотношении 75:25

Удаление столбца 'id' из датасетов

In [5]:
df_0.drop(columns = ['id'], inplace = True)
df_1.drop(columns = ['id'], inplace = True)
df_2.drop(columns = ['id'], inplace = True)

# проверка результата
print(df_0.info())
print(df_1.info())
print(df_0.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 100000 entries, 0 to 99999
Data columns (total 4 columns):
 #   Column   Non-Null Count   Dtype  
---  ------   --------------   -----  
 0   f0       100000 non-null  float64
 1   f1       100000 non-null  float64
 2   f2       100000 non-null  float64
 3   product  100000 non-null  float64
dtypes: float64(4)
memory usage: 3.1 MB
None
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 100000 entries, 0 to 99999
Data columns (total 4 columns):
 #   Column   Non-Null Count   Dtype  
---  ------   --------------   -----  
 0   f0       100000 non-null  float64
 1   f1       100000 non-null  float64
 2   f2       100000 non-null  float64
 3   product  100000 non-null  float64
dtypes: float64(4)
memory usage: 3.1 MB
None
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 100000 entries, 0 to 99999
Data columns (total 4 columns):
 #   Column   Non-Null Count   Dtype  
---  ------   --------------   -----  
 0   f0       100000 non-null  float6

Определение функции для разбиения данных

In [6]:
def data_split(df):
    
    # формирование выборок с обучающими признаками и целевым признаком.
    features = df.drop(['product'], axis=1)
    target = df['product']

    # разделение обучающей и валидационной выборки в пропорции 75:25.
    features_train, features_valid, target_train, target_valid =\
        train_test_split(
        features,target,
        test_size=0.25, random_state=12345)

    #вывод размеров выборок для проверки
    print(features_train.shape)
    print(features_valid.shape)
    print(target_train.shape)
    print(target_valid.shape)
    
    return features_train, features_valid, target_train, target_valid

Разбиение данных.

In [7]:
# Разбиение данных датасета 0
df_0_features_train, df_0_features_valid,\
df_0_target_train, df_0_target_valid = data_split(df_0)

# Разбиение данных датасета 1
df_1_features_train, df_1_features_valid,\
df_1_target_train, df_1_target_valid = data_split(df_1)

# Разбиение данных датасета 2
df_2_features_train, df_2_features_valid,\
df_2_target_train, df_2_target_valid = data_split(df_2)

(75000, 3)
(25000, 3)
(75000,)
(25000,)
(75000, 3)
(25000, 3)
(75000,)
(25000,)
(75000, 3)
(25000, 3)
(75000,)
(25000,)


## Обучение и проверка модели

Определение функции для обучения и проверки модели.

In [8]:
def model_learning(features_train,target_train,features_valid,target_valid):
    
    # используем модель линейной регрессии. 
    model = LinearRegression()
    
    # обучение модели на тренировочной выборке
    model.fit(features_train,target_train) 
    
    # получение предсказания модели на валидационной выборке
    predictions_valid = model.predict(features_valid) 
    
    #выравнивание индексов
    predictions_valid = pd.Series(predictions_valid, target_valid.index)

    # расчет метрики RMSE на валидационной выборке
    rmse = mean_squared_error(target_valid, predictions_valid)**0.5 
    
    # расчет среднего запаса предсказанного сырья
    mean_predicted = predictions_valid.mean()
    
    return predictions_valid, rmse, mean_predicted

Обучение моделей для всех регионов.

In [9]:
# обучение модели для региона 0.
df_0_predicted,df_0_rmse, df_0_mean_predicted  = model_learning(
        df_0_features_train,
        df_0_target_train,
        df_0_features_valid,
        df_0_target_valid)
print(
    "RMSE модели линейной регрессии на валидационной выборке для региона 0:", 
    df_0_rmse)
print(
    "Средний запас предсказанного сырья для региона 0:", 
    df_0_mean_predicted)

# обучение модели для региона 1.
df_1_predicted, df_1_rmse, df_1_mean_predicted = model_learning(
        df_1_features_train,
        df_1_target_train,
        df_1_features_valid,
        df_1_target_valid)
print(
    "RMSE модели линейной регрессии на валидационной выборке для региона 1:", 
    df_1_rmse)
print(
    "Средний запас предсказанного сырья для региона 1:", 
    df_1_mean_predicted)

# обучение модели для региона 2.
df_2_predicted, df_2_rmse, df_2_mean_predicted = model_learning(
        df_2_features_train,
        df_2_target_train,
        df_2_features_valid,
        df_2_target_valid)
print(
    "RMSE модели линейной регрессии на валидационной выборке для региона 2:", 
    df_2_rmse)
print(
    "Средний запас предсказанного сырья для региона 2:", 
    df_2_mean_predicted)

RMSE модели линейной регрессии на валидационной выборке для региона 0: 37.5794217150813
Средний запас предсказанного сырья для региона 0: 92.59256778438035
RMSE модели линейной регрессии на валидационной выборке для региона 1: 0.8930992867756168
Средний запас предсказанного сырья для региона 1: 68.72854689544602
RMSE модели линейной регрессии на валидационной выборке для региона 2: 40.02970873393434
Средний запас предсказанного сырья для региона 2: 94.96504596800489


**Вывод:** от региона к региону, чем выше предсказанный объем, тем выше RMSE, то есть хуже качество обучения модели, и меньше доверия к предсказанному результату. При этом регион 1, имея самый низкий средний запас предсказываемого сырья, тем не менее, имеет значительно более низкую ошибку модели, а для нас точность предсказаний крайне важна для определния скважин с максимальным запасом.

## Подготовка к расчёту прибыли

### Рассчет минимально достаточного объём сырья для безубыточной разработки новой скважины.

In [10]:
#расчет минимальной выручки на скважину:
min_revenue_per_well = DEV_BUDGET / BEST_QUANTITY

#расчет минимальной продуктивности скважины:
minimal_units_volume = min_revenue_per_well / UNIT_REVENUE

print('Достаточный объём для безубыточной разработки новой скважины:', minimal_units_volume)

Достаточный объём для безубыточной разработки новой скважины: 111.11111111111111


**Вывод:** на данном этапе исследования ни один регион по предсказываемой средней продуктивности не превышает 95 единиц, что ниже уровня безубытка в 111.  
Для поиска возможностей прибыльной добычи нужны дальнейшие исследования.

## Расчёт прибыли и рисков для каждого региона

### Создадим функцию для расчёта прибыли по выбранным скважинам и предсказаниям модели

In [11]:
def wells_profit(target,predicted): 
    
    # сортировка скважин на предсказанной модели
    predicted_sorted = pd.Series(predicted).sort_values(ascending=False)
    
    # выборка реальных значений 200 лучших скважин по индексам предсказаний
    target_sorted = target[predicted_sorted.index][:BEST_QUANTITY]

    # суммирование целевого значения объема
    product = target_sorted.sum()
 
    #расчет прогнозируемой прибыли региона
    profit=(product * UNIT_REVENUE) - DEV_BUDGET
   
    
    return product, profit


### Рассчитаем продуктивность и прибыль  по выбранным скважинам и предсказаниям модели

Расчет прогнозируемых объема сырья и прибыли лучших 200 скважин для региона 0 

In [12]:
df_0_volume_sum, df_0_profit = wells_profit(
    df_0_target_valid,
    df_0_predicted)
print('Прогнозируемый объем сырья в регионе в единицах продукта:', round(df_0_volume_sum,2))
print('Прогнозируемая прибыль в регионе в млрд.рублей:', round(df_0_profit / 10**9,2))

Прогнозируемый объем сырья в регионе в единицах продукта: 29601.84
Прогнозируемая прибыль в регионе в млрд.рублей: 3.32


Расчет прогнозируемых объема сырья и прибыли для региона 1 

In [13]:
df_1_volume_sum, df_1_profit = wells_profit(
    df_1_target_valid,
    df_1_predicted)
print('Прогнозируемый объем сырья в регионе в единицах продукта:', round(df_1_volume_sum,2))
print('Прогнозируемая прибыль в регионе в млрд.рублей:', round(df_1_profit / 10**9,2))

Прогнозируемый объем сырья в регионе в единицах продукта: 27589.08
Прогнозируемая прибыль в регионе в млрд.рублей: 2.42


Расчет прогнозируемых объема сырья и прибыли для региона 2 

In [14]:
df_2_volume_sum, df_2_profit = wells_profit(
    df_2_target_valid,
    df_2_predicted)
print('Прогнозируемый объем сырья в регионе в единицах продукта:', round(df_2_volume_sum,2))
print('Прогнозируемая прибыль в регионе в млрд.рублей:', round(df_2_profit / 10**9,2))

Прогнозируемый объем сырья в регионе в единицах продукта: 28245.22
Прогнозируемая прибыль в регионе в млрд.рублей: 2.71


**Вывод:** Исходя из расчетов прогнозируемых объема сырья и прибыли на основании обученной модели наиболее привлекателен регион 0. При этом регион 1 уже не так сильно уступает остальным по рассчитанным показателям, а точность модели в нем значительно выше.

### Создадим функцию для расчёта рисков и распределения прибыли техникой Bootstrap

In [15]:
def profit_bootstrap(target, predicted):
    
    random_state = np.random.RandomState(12345)
    profit_all = []
    volume_all = []
    
    
    for i in range(1000):
        
        # формирование в цикле выборок по методу Bootstrap
        target_subsample = target.sample(
            EXPLORATION_QUANTITY, replace = True, random_state = random_state)
        predicted_subsample = predicted[target_subsample.index]
        
        #расчет прибыли и продуктивности для каждой выборки
        volume, profit = wells_profit(target_subsample,predicted_subsample)
        profit_all.append(profit)
        volume_all.append(volume)

    profit_all = pd.Series(profit_all)
    volume_all = pd.Series(volume_all)
    
    # рассчитываем среднюю прибыль
    profit_mean = profit_all.mean()
    
    # рассчитываем средний объем сырья
    volume_mean = volume_all.mean()
    
    #рассчитываем доверительный интервал
    lower = profit_all.quantile(q=0.025)
    upper = profit_all.quantile(q=0.975)

    # вычисляем риск как долю количества отрицательных значений прибыли  в выборках 
    risk = profit_all[profit_all < 0].count()/profit_all.count()
    
    # выводим результаты на экран
    print('Средний объем сырья в регионе:', round(volume_mean, 2), 'единиц')
    print('Средняя прибыль в регионе:', round(profit_mean/10**9, 2), 'млрд.руб')
    print('Доверительный интервал от', round(lower/10**9, 2),'млрд.руб.', 'до', round(upper/10**9, 2),'млрд.руб')
    print('Риск убытков:', risk*100, '%')
    
    # проверяем соответствие уровня риска установленному критерию
    if risk*100 > LOSS_PROBABILITY:
        print('Уровень риска региона превышает требуемый уровень в', LOSS_PROBABILITY, '%')
    else:
        print('Уровень риска региона соответствует требуемому уровеню не более', LOSS_PROBABILITY, '%')

### Найдем среднюю прибыль, 95%-й доверительный интервал и риск убытков в трех регионах

**Расчет для региона 0**

In [16]:
profit_bootstrap(df_0_target_valid, df_0_predicted)

Средний объем сырья в регионе: 23168.75 единиц
Средняя прибыль в регионе: 0.43 млрд.руб
Доверительный интервал от -0.1 млрд.руб. до 0.95 млрд.руб
Риск убытков: 6.0 %
Уровень риска региона превышает требуемый уровень в 2.5 %


**Расчет для региона 1**

In [17]:
profit_bootstrap(df_1_target_valid,df_1_predicted)

Средний объем сырья в регионе: 23367.16 единиц
Средняя прибыль в регионе: 0.52 млрд.руб
Доверительный интервал от 0.07 млрд.руб. до 0.93 млрд.руб
Риск убытков: 1.0 %
Уровень риска региона соответствует требуемому уровеню не более 2.5 %


**Расчет для региона 2**

In [18]:
profit_bootstrap(df_2_target_valid, df_2_predicted)

Средний объем сырья в регионе: 23188.91 единиц
Средняя прибыль в регионе: 0.44 млрд.руб
Доверительный интервал от -0.13 млрд.руб. до 0.97 млрд.руб
Риск убытков: 6.4 %
Уровень риска региона превышает требуемый уровень в 2.5 %


## Общие выводы

**По проекту было выполнено следующее:**

- Импорт и обзор данных
- Обработка данных: 
   - проверка на наличие пропусков в данных
   - проверка на наличие явных дубликатов
   - удаление лишних данных
   - разбиение данных на обучающую и валидационную выборки в пропорции 75:25
- Обучение и проверка модели для каждого региона:
 - обучение модели и получение предсказаний на валидационной выборке
 - расчет среднего запаса предсказанного сырья и RMSE модели.
- Подготовка к расчёту прибыли:
 - создание переменных с ключевыми значениями для расчётов
 - рассчет достаточного объёма сырья для безубыточной разработки новой скважины
- Рассчет прибыли и рисков для каждого региона:
 - рассчет продуктивности и прибыли по лучшим 200 выбранным скважинам и предсказаниям модели для каждого региона
 - рассчет с применением метода Bootstrap следующих показателей по каждому региону:
   - средний объем сырья
   - средняя прибыль
   - доверительный интервал для прибыли
   - риск убытков

**Выводы:**
- Из трех регионов только регион 1 (geo_data_1) показал уровень риска меньше заданого критерия 2.5%, и соответственно он один должен быть ототбран.
- Также из трех регионов именно регион 1 показал наибольшую среднюю прибыль.
- Кроме того, в данном регионе 1 наблюдается самый низкий доверительный интервал.
- Также следует учесть, что в регионе 1 наблюдается самая высокая точность предсказаний модели 

**Рекомендации сотрудникам нефтедобывающей компании:** 
- На основании построенной в рамках данного проекта модели рекомендуется выбрать регион 1. 